### PR TITLE
Phase 5: Low FPS Test Coverage and Canary Stability (squashed)

### PR DESCRIPTION
**Pull request by YPandas**
_What is changed_

- Added unified `storage_runner.groovy` that consolidates all cron-triggered storage canary scenarios into a single runner
- Migrated all storage jobs from self-rescheduling pipelines to cron-triggered execution (`prod_storage_cron.txt`)
- Removed `gamma_orchestrator.groovy` and `gamma_seed.groovy` — gamma is now cron-based like prod
- Removed all storage code from `runner.groovy` and all storage jobs from `orchestrator.groovy` (full decoupling)
- Added `StorageLowFps` canary test (10fps) to validate storage and viewer behavior with reduced media throughput
- Added `.in_use` lockfile protection to prevent cron cleanup scripts from deleting active Jenkins workspaces mid-build
- Added `lock(resource: RUNNER_LABEL)` in gamma runner to prevent concurrent builds of the same test type
- Added cleanup script auto-sync from repo when code changes (master, viewer, consumer nodes)
- Added `failFast` interrupt handling — `FlowInterruptedException` caught and marked UNSTABLE instead of ABORTED
- Added `PipelineKeepAlive` checkpoint stages for pipeline observability (GitCheckoutComplete, MasterStarted, MasterFinished, etc.)
- Added `MasterStreamingAvailability` heartbeat metric (every 20s) and `ViewerStreamingAvailability` heartbeat metric
- Added JoinSS-anchored timing metrics for both C Master and JS Viewer (TimeToSendIce, TimeToReceiveIce, TimeToPeerConnection, JoinSSToFirstFrame)
- Added periodic RTP/ICE stats collection for both C Master and JS Viewer (RoundTripTime, JitterBufferDelay, DecodeTime, bitrate, packet rates)
- Added `ViewerHttpError` metric to separate HTTP errors from JoinSSAsViewerAvailability
- Refined `PeerConnectionAvailability` to be state-based with scoped disconnect tracking
- Added `build-storage-master.sh` for persistent builds with skip-rebuild logic
- Added `alarm-sop.md` documentation for canary metric investigation
- Replaced opus sample frames with correctly extracted 7798 frames from source MP4 (4-digit naming)
- Added `audio-source.wav` asset for headless Chrome audio input (fake audio capture)
- Added C++ metrics: `MasterStreamingAvailability`, `TimeToSendIce`, `TimeToReceiveIce`, `JoinSSCallToFirstFrame`, `RoundTripTime`, periodic RTP outbound stats (PacketsSentPerSecond, OutgoingBitrate, PliCountPerSecond)
- Fixed git checkout for force-pushed branches (`reset --hard` to origin after fetch)

_Why it's changed_

The self-rescheduling pipeline model caused unbounded Jenkins build queue growth — each completed run triggered the next, and if builds backed up (node busy, network issue), the queue would grow indefinitely. Cron-based execution guarantees fixed frequency and eliminates queue bloat.

The `.in_use` lockfile and `lock(resource:)` solve two recurring stability issues: cron cleanup scripts deleting workspaces mid-build (causing mysterious git/npm failures), and duplicate concurrent builds racing on the same scenario when cron fires while a previous run is still completing.

The new metrics (streaming availability heartbeats, JoinSS-anchored timing, periodic RTP stats) fill observability gaps where previously we only had point-in-time measurements at connection establishment. Now we have continuous signals during the streaming session itself, enabling detection of mid-session degradation that connection-level metrics miss entirely.

The low FPS test validates that the storage pipeline and viewer correctly handle reduced frame rates — a scenario that exposed buffering and timeout edge cases in both the media server and the JS SDK.

_How it's changed_

- **New files**: `jobs/storage_runner.groovy` (unified cron-triggered runner, 922 lines), `jobs/cron/prod_storage_cron.txt` (8 cron entries for all prod scenarios), `scripts/build-storage-master.sh` (persistent build with skip-rebuild), `scripts/cron/cleanup-master.sh`, `scripts/cron/cleanup-viewer.sh`, `scripts/cron/cleanup-consumer.sh`, `docs/alarm-sop.md`, `assets/audio-source.wav`
- **New assets**: 7798 Opus sample frames (`assets/opusSampleFrames/sample-0000.opus` through `sample-7797.opus`) — correctly extracted from source MP4 with 4-digit naming
- **Removed**: `jobs/gamma_orchestrator.groovy`, `jobs/gamma_seed.groovy`
- **`runner.groovy`**: Removed all storage stages and logic (720→~80 lines of storage code removed); now exclusively handles P2P canary
- **`orchestrator.groovy`**: Removed all storage job definitions (cron owns the lifecycle)
- **`gamma_runner.groovy`**: Reworked to cron-triggered model; added `lock(resource:)`, failFast interrupt handling, `.in_use` lockfile, cleanup script sync, `PipelineKeepAlive` checkpoints
- **`chrome-headless.js`**: Added `ViewerStreamingAvailability` heartbeat (20s), JoinSS-anchored timing metrics (TimeToReceiveIce, TimeToSendIce, TimeToPeerConnection, JoinSSToFirstFrame), periodic RTP inbound stats (RoundTripTime, JitterBufferDelay, DecodeTime, NackCountPerSecond, PliCountPerSecond, FramesDecodedPerSecond, FramesDroppedPerSecond, IncomingBitrateKbps, PacketsReceivedPerSecond, PacketsLostPerSecond), end-of-session totals, `ViewerHttpError` separation, refined `PeerConnectionAvailability` logic
- **`prepare-storage-viewer.sh`**: Major rework for cron-based model; added JS SDK branch checkout with webpack dev server, persistent venv, skip-rebuild
- **C++ source** (`CloudwatchMonitoring.cpp/h`): Added `pushMasterStreamingAvailability`, `pushTimeToPeerConnection`, `pushTimeToSendAnswer`, `pushTimeToReceiveInboundMedia`, `pushTimeToSendIce`, `pushTimeToReceiveIce`, `pushJoinSSCallToFirstFrame`, `pushRoundTripTime`, periodic RTP stats (PacketsSentPerSecond, PacketsReceivedPerSecond, OutgoingBitrate, PliCountPerSecond), end-of-session totals
- **C++ source** (`Common.cpp`): Added `canaryMasterStreamingAvailability` timer callback (20s heartbeat), periodic ICE candidate pair stats extraction (RTT, packets, bitrate), PLI rate computation
- **C++ source** (`kvsWebRTCClientMaster.cpp`): Added JoinSS timestamp capture, timing metric pushes at connection/ICE/first-frame milestones, video-only mode support (`CANARY_MEDIA_TYPE`), no-loop mode (`NO_LOOP_FRAMES`)
- **C++ source** (`Samples.h`, `Peer.h`, `Include.h`): Added `rtcMetricsHistory` struct, `NUMBER_OF_OPUS_FRAME_FILES` updated to 7798, `noLoopFrames` flag

**New metrics:**

| Metric | Namespace | Type | Description |
|--------|-----------|------|-------------|
| `MasterStreamingAvailability` | KinesisVideoSDKCanary | Count (0/1) | Heartbeat every 20s: 1 while CONNECTED, 0 on FAILED/DISCONNECTED |
| `ViewerStreamingAvailability` | ViewerApplication | Count (0/1) | Heartbeat every 20s: 1 while storageSessionActive, 0 on drop |
| `TimeToReceiveIce` (Viewer) | ViewerApplication | Milliseconds | JoinSS API call to first ICE candidate received |
| `TimeToSendIce` (Viewer) | ViewerApplication | Milliseconds | JoinSS API call to first ICE candidate sent |
| `TimeToPeerConnection` (Viewer) | ViewerApplication | Milliseconds | JoinSS API call to peer connection established |
| `JoinSSToFirstFrame` (Viewer) | ViewerApplication | Milliseconds | JoinSS API call to first video frame received |
| `TimeToSendIce` (Master) | KinesisVideoSDKCanary | Milliseconds | JoinSS API call to first ICE candidate sent |
| `TimeToReceiveIce` (Master) | KinesisVideoSDKCanary | Milliseconds | JoinSS API call to first ICE candidate received |
| `JoinSSCallToFirstFrame` (Master) | KinesisVideoSDKCanary | Milliseconds | JoinSS API call to first frame sent |
| `RoundTripTime` | KinesisVideoSDKCanary / ViewerApplication | Milliseconds | ICE candidate pair round-trip time |
| `JitterBufferDelay` | ViewerApplication | Milliseconds | Jitter buffer delay from RTCInboundRtpStreamStats |
| `DecodeTime` | ViewerApplication | Milliseconds | Frame decode time from RTCInboundRtpStreamStats |
| `NackCountPerSecond` | ViewerApplication | Count/sec | NACK requests per second (periodic) |
| `PliCountPerSecond` | ViewerApplication / KinesisVideoSDKCanary | Count/sec | PLI requests per second (periodic) |
| `FramesDecodedPerSecond` | ViewerApplication | Count/sec | Frames decoded per second (periodic) |
| `FramesDroppedPerSecond` | ViewerApplication | Count/sec | Frames dropped per second (periodic) |
| `IncomingBitrateKbps` | ViewerApplication | Kbps | Incoming bitrate (periodic) |
| `PacketsReceivedPerSecond` | ViewerApplication | Count/sec | Packets received per second (periodic) |
| `PacketsLostPerSecond` | ViewerApplication | Count/sec | Packets lost per second (periodic) |
| `TotalPacketsReceived` | ViewerApplication | Count | Cumulative packets received (end-of-session) |
| `TotalFramesDecoded` | ViewerApplication | Count | Cumulative frames decoded (end-of-session) |
| `TotalFramesDroppedPercentage` | ViewerApplication | Percent | framesDropped / (decoded + dropped) × 100 (end-of-session) |
| `ViewerHttpError` | ViewerApplication | Count (0/1) | 1 on HTTP error during JoinStorageSession API call |

**PipelineKeepAlive checkpoint stages:**

| Metric | Emitted After |
|--------|---------------|
| `GitCheckoutComplete` | Git repo fetched and checked out on master node |
| `MasterStarted` | C Master binary launched (streaming begins) |
| `MasterFinished` | C Master binary exited (streaming ended) |
| `ViewerPrepareComplete` | Viewer dependencies installed and ready |
| `ViewerSessionComplete` | Viewer session finished (all cycles done) |
| `CleanupComplete` | Workspace cleanup and .in_use removal done |

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp [hunk 1]
```diff
@@ -221,15 +221,47 @@ PVOID sendVideoPackets(PVOID args)
...
+        videoFrameDuration = SAMPLE_VIDEO_FRAME_DURATION;
+        DLOGI("[KVS Master] No CANARY_FRAME_RATE set, using default %u fps", DEFAULT_FPS_VALUE);
+    }
+
     frame.presentationTs = 0;
     startTime = GETTIME();
     lastFrameTime = startTime;
 
+    // Check if we should stop after one pass through all frames (no looping)
+    pNoLoopFrames = GETENV("CANARY_NO_LOOP_FRAMES");
```
Comment (sirknightj) [inline canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp:252]: Is this documented somewhere?

Comment (sirknightj) [reply canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp:252]: Consider: keeping all the configuration parameters close to each other physically

File: canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp [hunk 2]
```diff
@@ -266,6 +298,14 @@ PVOID sendVideoPackets(PVOID args)
                 DLOGD("[Canary] Start up latency from offer received to first frame sent (timeToFirstFrame): %lf ms", timeToFirstFrame);
                 Canary::Cloudwatch::getInstance().monitoring.pushTimeToFirstFrame(timeToFirstFrame,
                                                                                 Aws::CloudWatch::Model::StandardUnit::Milliseconds);
+
+                // Push JoinSSCallToFirstFrame — time from JoinStorageSession call to first frame sent
```
Comment (sirknightj) [inline canary/webrtc-c/src/media-server-storage/kvsWebRTCClientMaster.cpp:302]: first frame sent _successfully_ (if statement is in a status==success), after the peer connectionand srtp/dtls is established

File: canary/webrtc-c/src/CloudwatchMonitoring.h [hunk 3]
```diff
@@ -31,6 +31,24 @@ class CloudwatchMonitoring {
...
+    VOID pushTimeToSendIce(UINT64, Aws::CloudWatch::Model::StandardUnit);
+    VOID pushTimeToReceiveIce(UINT64, Aws::CloudWatch::Model::StandardUnit);
+    VOID pushJoinSSCallToFirstFrame(UINT64, Aws::CloudWatch::Model::StandardUnit);
+    VOID pushRoundTripTime(DOUBLE, Aws::CloudWatch::Model::StandardUnit);
+
+    // Periodic during-session metrics
+    VOID pushPacketsSentPerSecond(DOUBLE);
+    VOID pushPacketsReceivedPerSecond(DOUBLE);
+    VOID pushOutgoingBitrate(DOUBLE);
+    VOID pushPliCountPerSecond(DOUBLE);
```
Comment (sirknightj) [inline canary/webrtc-c/src/CloudwatchMonitoring.h:44]: Is this tracking PLI received from kvs storage peer?

**Review by sirknightj — APPROVED** [review]

**Review by vabhasin — APPROVED** [review]
